Add runtime-check replacing instrumentation pass

xbauch opened the pull request:
replacing calls to `__CPROVER_X_check-name` checks with a user-specified
combination of assume/assert (for a user-defined language X).

For example Ada checks can be generated with `gnat2goto`, and this pass allows
us to have more fine-grained and option-able workflow when verifying them. Also this allows us to customise how `check-name` is reported to the user via the assertion description.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

xbauch commented:
@martin-cs  Hi, Martin. This PR is supposed to be the second part of https://github.com/martin-cs/ASVAT/issues/9. What do you think?

hannes-steffenhagen-diffblue commented on src/goto-instrument/ada_checks.cpp:
```diff
@@ -0,0 +1,182 @@
+private:
+  goto_modelt &goto_model;
+
+  enum class Replace_Type
```
:pick: `replace_typet`

hannes-steffenhagen-diffblue commented on src/goto-instrument/ada_checks.cpp:
```diff
@@ -0,0 +1,182 @@
+
+  enum class Replace_Type
+  {
+    AssertOnly,
```
:pick: `assert_only`, `assume_only`, `assert_then_assume`

hannes-steffenhagen-diffblue commented on src/goto-instrument/ada_checks.cpp:
```diff
@@ -0,0 +1,182 @@
+    AssertThenAssume
+  };
+
+  enum class Ada_Check
```
naming etc

xbauch replied:
Renamed. Here and above.

hannes-steffenhagen-diffblue commented on src/goto-instrument/ada_checks.cpp:
```diff
@@ -0,0 +1,182 @@
+  Ada_Check get_check_type(const std::string &function_name) const
+  {
+    if(function_name.find("Overflow") != std::string::npos)
+      return Ada_Check::Overflow;
```
Hm, shouldn't these names be fixed? If that's the case, shouldn't these be exact matches?

xbauch replied:
Sure, why not. Changed.

hannes-steffenhagen-diffblue commented on src/goto-instrument/ada_checks.cpp:
```diff
@@ -0,0 +1,182 @@
+
+  Replace_Type get_replace_type(const std::string &replace_option) const
+  {
+    if(replace_option == "assert-first-arg")
```
What's with the `first-arg`?

xbauch replied:
https://github.com/martin-cs/ASVAT/issues/9 .. That is the exact wording is part of the task specification.

hannes-steffenhagen-diffblue commented on src/goto-instrument/ada_checks.cpp:
```diff
@@ -0,0 +1,182 @@
+      "unknown Ada check replacement option" + replace_option, "--ada-checks"};
+  }
+
+  std::string type2string(Ada_Check type) const
```
`Ada_Check` type2string if anything I guess? I think you could also just overload `std::to_string` for this

xbauch replied:
I used the `to_string`.

hannes-steffenhagen-diffblue commented on src/goto-instrument/ada_checks.cpp:
```diff
@@ -0,0 +1,182 @@
+    case Ada_Check::DivByZero:
+      return "Division by zero";
+    }
+    UNREACHABLE;
```
Funnily enough, this is actually very easily reachable, you just have to call it with stupid arguments (no need to change this, just my attempt at a funny remark)

xbauch replied:
Even if the argument is an `enum class`?

hannes-steffenhagen-diffblue replied:
Yes, hilariously enough. `std::byte` is defined without any named constants but is used to hold any byte value (however, I think treating an enum that isn't explicitly intended to be used as a bitfield or a numeric type like that shouldn't pass code review, hence why this is just a passing remark rather than a genuine warning)

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 1b87ab2c65bce47548d8489752a4f316f5409da6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/128026883

NlightNFotis commented on src/goto-instrument/ada_checks.cpp:
```diff
@@ -0,0 +1,182 @@
+  const std::string &replace_option)
+{
+  exprt boolean_condition = condition;
+  while(boolean_condition.id() == ID_typecast)
```
Why is this `while`? I thought the `to_typecast_expr` below always simplifies? Is it possible it returns something that has the same structure, and hasn't been structurally reduced?

NlightNFotis replied:
Okay, ignore this comment. After I wrote it, I remembered that the `to_typecast_expr` is not even simplifying, it's just a cast. But I still wonder why the need for `while` here.

xbauch replied:
In a C program this is probably unnecessary, but gnat2goto can produce nested typecasts.

NlightNFotis commented on src/goto-instrument/ada_checks.cpp:
```diff
@@ -0,0 +1,182 @@
+  }
+
+  auto skip_instruction = goto_programt::make_skip();
+  instruction->swap(skip_instruction);
```
Why is this happening here? 

(I'm not saying it's wrong, I just don't understand what purpose it serves)

xbauch replied:
Well the idea is that we _replace_ the original call to `__CPROVER_Ada_check` with assert/assume. On the lines above we have added the assertion, and here we change the original function call to a `SKIP`. The only interface for changing an existing instruction I could find is `swap`, so that's what I used.

NlightNFotis reviewed (commented):
It looks good to me, I just have two questions:

hannes-steffenhagen-diffblue commented on src/goto-instrument/ada_checks.cpp:
```diff
@@ -0,0 +1,182 @@
+  const std::string &replace_option)
+{
+  ada_checkst ada_checks{goto_model};
+  ada_checks.replace_with(replace_option);
```
FYI I think you need to call `update` somewhere to fix the location numbers after inserting or removing instructions, unless that was changed (or maybe that's already done by goto-instrument? No idea tbh)

xbauch replied:
I added the updating, once all check are replaced.

hannes-steffenhagen-diffblue reviewed (approved):
Some nitpicks, I think the naming needs to be changed, otherwise LGTM

xbauch commented on regression/goto-instrument/ada-checks/test.desc:
```diff
@@ -0,0 +1,9 @@
+--ada-checks assert-then-assume-first-arg
+^EXIT=10$
+^SIGNAL=0$
+^\[1\] Ada Overflow check: FAILURE$
```
The goto-functions line looks like this:
`ASSERT i > 5 // Ada Overflow check`
but somehow the asserted condition is not included in the CBMC output.

xbauch replied:
Fixed.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: d1ecf617715a82dc8dfed6f09c4b0e6d672b7963).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/128144409

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7c23fd75c2953de1aa4ca13829c214b8dc2197bf).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/128161589

petr-bauch committed:
Add runtime-check replacing instrumentation pass

replacing calls to `__CPROVER_X_check-name` checks with a user-specified
combination of assume/assert (for a user-defined language X).

For example Ada checks can be generated with `gnat2goto`, and this pass allows
us to have more fine-grained and option-able workflow when verifying them.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 94b39beb46e219f383d86a9dbe4f18c7ba683343).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/128366180

martin-cs commented on regression/goto-instrument/ada-checks/main.c:
```diff
@@ -0,0 +1,10 @@
+void __CPROVER_Ada_Overflow_Check(int condition)
+{
+}
```
I think it might be nice to have the typical assert-then-assume body here.

martin-cs reviewed (commented):
I had been thinking of using `--generate-function-body` and one of the inlining functions.  I think this would simplify things a bit.

xbauch commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -1759,6 +1771,16 @@ void goto_instrument_parse_optionst::help()
+    // NOLINTNEXTLINE(whitespace/line_length)
+    "                              <rplc-option> should be \'assert-first-arg\'\n"
+    // NOLINTNEXTLINE(whitespace/line_length)
+    "                              \'assert-first-arg\' or \'assert-then-assume-first-arg\'\n"
```
`assume-first-arg`

petr-bauch committed:
Alternative implementation via generate-function-body

the idea is then to inline the function call.

xbauch commented:
@martin-cs The second commit tries to do the same using `generate-function-body` pass. There are several problems with that: 
1) the `generate-function-body` didn't like generating bodies for function that already had bodies
2) the `--function-inline foo` option inlines all calls to functions inside `foo` (and not the calls to `foo`), we therefore have to use `--function-inline main`
3) inlining creates temporary variables for arguments and thus the original asserted expressions are not visible in the CBMC report. The `main` looks like this:
```
main /* main */
        // 0 file main.c line 7 function main
        signed int i;
        // 1 file main.c line 7 function main
        i = 0;
        // 2 file main.c line 8 function main
        LOCATION
        // 3 file main.c line 8 function main
        signed int condition;
        // 4 file main.c line 8 function main
        condition = (signed int)(i > 5);
        // 5 file main.c line 8 function main
        ASSERT (__CPROVER_bool)condition // assertion cast(__CPROVER_Ada_Overflow_Check::condition, 𝔹)
        // 6 file main.c line 8 function main
        ASSUME (__CPROVER_bool)condition
        // 7 file main.c line 8 function main
```
Otherwise it works.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 1def9093cbbe90fecf3458cb25c5a5b282ffd670).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/128520921

martin-cs commented:
> 1. the generate-function-body didn't like generating bodies for function that already had bodies

It should be easy enough to fix this by deleting the bodies first.


> 2. the --function-inline foo option inlines all calls to functions inside foo (and not the calls to foo), 
> we therefore have to use --function-inline main

Ah; sorry.  There are two ways of inlining things, either inline everything *in* the target function or inline every call *to* the target function.  Unfortunately CPROVER in both code and options is remarkably unclear which is which and it looks like I picked the wrong one.

My suggestion was to inline every call *to* the run-time checks.  Inlining everything into `main` won't scale and is kinda scary.

> 3. inlining creates temporary variables for arguments and thus the original asserted expressions 
> are not visible in the CBMC report.

That is annoying.  If the other set of inlining function have this as well then we may have to work around this.

martin-cs commented on src/goto-instrument/replace_runtime_checks.cpp:
```diff
@@ -0,0 +1,209 @@
+    overflow,
+    range,
+    division_by_zero
+  };
```
Do we need to enumerate these here?  IIRC there are a number of others and it would be a little bit annoying to have to do parallel maintenance between this and gnat2goto.

martin-cs commented on src/goto-instrument/replace_runtime_checks.cpp:
```diff
@@ -0,0 +1,209 @@
+    if(function_name == common_prefix + "_Range_Check")
+      return runtime_check_typet::range;
+    if(function_name == common_prefix + "_Division_Check")
+      return runtime_check_typet::division_by_zero;
```
Is there a way of doing "all checks"

martin-cs reviewed (approved):
I think this is a good implementation of this design.  Please have another look and see if the generate body / inline calls option works and produces something more compact.  If not then I am fine for this to be merged.

xbauch commented:
> My suggestion was to inline every call _to_ the run-time checks. Inlining everything into `main` won't scale and is kinda scary.

If there is such an option, I did not find it.

xbauch commented:
@martin-cs How do you want to handle this one? I did not find a version of `function-inline` that would inline all calls to a target function. I guess we could write our own.

petr-bauch committed:
Implement inliner that works per-function

rather then the default which inline all calls in a function.

xbauch commented:
@martin-cs So I've implemented an inliner that inlines calls to a function, but if we want it to preserve argument values (so that we can see the assertions in the cbmc report), that would be a much bigger change. For that, the version in the first commit would be easier.

TGWDB commented:
Closing due to age and "Draft" status with lack of progress/update in years. Please reopen if you will update this and progress upon it.
